lib/semian/grpc.rb

```diff
@@ -0,0 +1,94 @@
+require 'semian/adapter'
+require 'grpc'
+
+module GRPC
+  GRPC::Unavailable.include(::Semian::AdapterError)
+
+  class SemianError < GRPC::BadStatus
+    attr_reader :details
+
+    def initialize(semian_identifier, *args)
+      super(*args)
+      @details = message
+      @semian_identifier = semian_identifier
+    end
+  end
+
+  ResourceBusyError = Class.new(SemianError)
+  CircuitOpenError = Class.new(SemianError)
+end
+
+module Semian
+  module GRPC
+    attr_reader :raw_semian_options
+    include Semian::Adapter
+
+    ResourceBusyError = ::GRPC::ResourceBusyError
+    CircuitOpenError = ::GRPC::CircuitOpenError
+
+    class SemianConfigurationChangedError < RuntimeError
+      def initialize(msg = "Cannot re-initialize semian_configuration")
+        super
+      end
+    end
+
+    class << self
+      attr_accessor :exceptions
+      attr_reader :semian_configuration
+
+      def semian_configuration=(configuration)
+        raise Semian::GRPC::SemianConfigurationChangedError unless @semian_configuration.nil?
+        @semian_configuration = configuration
+      end
+
+      def retrieve_semian_configuration(host)
+        @semian_configuration.call(host) if @semian_configuration.respond_to?(:call)
+      end
+    end
+
+    def raw_semian_options
+      @raw_semian_options ||= begin
+        # If the host is empty, it's possible that the adapter was initialized
+        # with the channel. Therefore, we look into the channel to find the host
+        if @host.empty?
+          host = @ch.target
+        else
+          host = @host
+        end
+        @raw_semian_options = Semian::GRPC.retrieve_semian_configuration(host)
+        @raw_semian_options = @raw_semian_options.dup unless @raw_semian_options.nil?
+      end
+    end
+
+    def semian_identifier
+      @semian_identifier ||= raw_semian_options[:name]
+    end
+
+    def resource_exceptions
+      [
+        ::GRPC::DeadlineExceeded,
+        ::GRPC::ResourceExhausted,
+        ::GRPC::Unavailable,
+        ::GRPC::BadStatus,
```

Comment:
Since `BadStatus` is the parent of all GRPC errors, we likely don't want to list it here.

This list should only contain "transient" errors, e.g. errors we have reason to believe are due to general bad health of the service, and are likely to be resolved later on.

So this shouldn't include things like validation errors, bad requests, authentication error, etc.

Additionally, ideally all classes in the list should include `Semian::AdapterError`. It's not absolutely mandatory, but it's useful to distinguish which backend is actually in bad health.

Reply:
I'll take a deeper look on each error. Thanks!

Reply:
Finally found the right documentation for errors: https://grpc.io/grpc/core/md_doc_statuscodes.html

Reply:
For reference, there's more details about each errors here: https://grpc.io/grpc/core/md_doc_statuscodes.html